File: iOS/Kalorie/Core/Models/MyCreatedMealModel.swift
//
//  MyCreatedMealModel.swift
//  Kalorie
//
//  Created by Josef Antoni on 20.08.2026.
//

import Foundation
import MacroKit

struct MyCreatedMealIngredientDomain: Equatable {

    // MARK: - Properties

    let foodItemId: String
    let czName: String
    let engName: String
    let grams: Double
    let nutrition: FoodNutritionValues
}

struct MyCreatedMealDomain {

    // MARK: - Properties

    let id: String
    let name: String
    let ingredients: [MyCreatedMealIngredientDomain]
    let createdAt: Date
    let updatedAt: Date
    let portions: [FoodPortionDomain]

    // MARK: - Init

    init(
        id: String,
        name: String,
        ingredients: [MyCreatedMealIngredientDomain],
        createdAt: Date,
        updatedAt: Date,
        portions: [FoodPortionDomain] = []
    ) {
        self.id = id
        self.name = name
        self.ingredients = ingredients
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.portions = portions
    }
}

extension MyCreatedMealDomain {

    // MARK: - Functions

    func asFoodItem() -> FoodItemDomain {
        let gramsList = ingredients.map(\.grams)
        let totalGrams = gramsList.reduce(0, +)
        func density(_ value: (FoodNutritionValues) -> Double) -> Double {
            weightedMeanPerHundredGrams(values: ingredients.map { value($0.nutrition) }, grams: gramsList)
        }
        func densityOptional(_ value: (FoodNutritionValues) -> Double?) -> Double? {
            var values: [Double] = []
            for ingredient in ingredients {
                guard let resolved = value(ingredient.nutrition) else { return nil }
                values.append(resolved)
            }
            return weightedMeanPerHundredGrams(values: values, grams: gramsList)
        }
        return FoodItemDomain(
            id: id,
            kind: .createdMeal,
            czName: name,
            engName: "",
            weight: totalGrams,
            date: createdAt,
            energyKJ: density(\.energyKJ),
            caloriesPerHundredGrams: density(\.caloriesPerHundredGrams),
            fat: density(\.fat),
            fatSaturated: densityOptional(\.fatSaturated),
            fatUnsaturatedFattyAcids: density(\.fatUnsaturatedFattyAcids),
            carbohydrate: density(\.carbohydrate),
            carbohydratePureSugar: density(\.carbohydratePureSugar),
            fiber: densityOptional(\.fiber),
            protein: density(\.protein),
            salt: density(\.salt),
            portions: portions
        )
    }
}

private func weightedMeanPerHundredGrams(values: [Double], grams: [Double]) -> Double {
    MacrosKt.weightedMeanPerHundredGrams(
        values: values.map { KotlinDouble(double: $0) },
        grams: grams.map { KotlinDouble(double: $0) }
    )
}

enum MyCreatedMealError: Error {
    case invalidName
    case noIngredients
    case invalidIngredientWeight
    case invalidPortion
}

enum MyCreatedMealValidation {

    // MARK: - Functions

    static func validate(
        name: String,
        ingredients: [MyCreatedMealIngredientDomain],
        portions: [FoodPortionDomain] = []
    ) -> MyCreatedMealError? {
        let trimmedName = name.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmedName.isEmpty { return .invalidName }
        if ingredients.isEmpty { return .noIngredients }
        if ingredients.contains(where: { $0.grams < 1 }) { return .invalidIngredientWeight }
        if portions.contains(where: { FoodPortionValidation.validate(name: $0.name, grams: $0.grams) != nil }) { return .invalidPortion }
        if FoodPortionValidation.validate(portions: portions) != nil { return .invalidPortion }
        return nil
    }

    static func canSave(
        name: String,
        ingredients: [MyCreatedMealIngredientDomain],
        portions: [FoodPortionDomain] = []
    ) -> Bool {
        validate(name: name, ingredients: ingredients, portions: portions) == nil
    }
}

File: Android/app/src/main/java/antoni/kalorie/core/models/MyCreatedMealModel.kt
package antoni.kalorie.core.models

import antoni.kalorie.macrokit.weightedMeanPerHundredGrams
import java.time.Instant

data class MyCreatedMealIngredientDomain(
    val foodItemId: String,
    val czName: String,
    val engName: String,
    val grams: Double,
    val nutrition: FoodNutritionValues,
)

data class MyCreatedMealDomain(
    val id: String,
    val name: String,
    val ingredients: List<MyCreatedMealIngredientDomain>,
    val createdAt: Instant,
    val updatedAt: Instant,
    val portions: List<FoodPortionDomain> = emptyList(),
) {

    // MARK: - Functions

    fun asFoodItem(): FoodItemDomain {
        val gramsList = ingredients.map { it.grams }
        val totalGrams = gramsList.sum()
        fun density(value: (FoodNutritionValues) -> Double): Double =
            weightedMeanPerHundredGrams(values = ingredients.map { value(it.nutrition) }, grams = gramsList)
        fun densityOptional(value: (FoodNutritionValues) -> Double?): Double? {
            val values = ingredients.map { value(it.nutrition) ?: return null }
            return weightedMeanPerHundredGrams(values = values, grams = gramsList)
        }
        return FoodItemDomain(
            id = id,
            kind = FoodItemKind.CREATED_MEAL,
            czName = name,
            engName = "",
            weight = totalGrams,
            date = createdAt,
            energyKJ = density { it.energyKJ },
            caloriesPerHundredGrams = density { it.caloriesPerHundredGrams },
            fat = density { it.fat },
            fatSaturated = densityOptional { it.fatSaturated },
            fatUnsaturatedFattyAcids = density { it.fatUnsaturatedFattyAcids },
            carbohydrate = density { it.carbohydrate },
            carbohydratePureSugar = density { it.carbohydratePureSugar },
            fiber = densityOptional { it.fiber },
            protein = density { it.protein },
            salt = density { it.salt },
            portions = portions,
        )
    }
}

sealed class MyCreatedMealError : Exception() {
    data object InvalidName : MyCreatedMealError()
    data object NoIngredients : MyCreatedMealError()
    data object InvalidIngredientWeight : MyCreatedMealError()
    data object InvalidPortion : MyCreatedMealError()
}

object MyCreatedMealValidation {

    // MARK: - Functions

    fun validate(
        name: String,
        ingredients: List<MyCreatedMealIngredientDomain>,
        portions: List<FoodPortionDomain> = emptyList(),
    ): MyCreatedMealError? {
        if (name.isBlank()) return MyCreatedMealError.InvalidName
        if (ingredients.isEmpty()) return MyCreatedMealError.NoIngredients
        if (ingredients.any { it.grams < 1 }) return MyCreatedMealError.InvalidIngredientWeight
        if (portions.any { FoodPortionValidation.validate(name = it.name, grams = it.grams) != null }) return MyCreatedMealError.InvalidPortion
        if (FoodPortionValidation.validate(portions) != null) return MyCreatedMealError.InvalidPortion
        return null
    }

    fun canSave(
        name: String,
        ingredients: List<MyCreatedMealIngredientDomain>,
        portions: List<FoodPortionDomain> = emptyList(),
    ): Boolean = validate(name = name, ingredients = ingredients, portions = portions) == null
}
